keps/sig-node/3673-kubelet-parallel-image-pull-limit/README.md

```diff
@@ -0,0 +1,856 @@
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This KEP proposes adding to kubelet a node-level limit on the number of parallel image pulls.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### QPS/burst limits on kubelet are confusing
+Currently kubelet limits image pulls with QPS and burst, but they are confusing as they only limit the number of requests sent to container runtime, and does not consider the number of inflight image pulls. In other words, even a small QPS is set, there still could be many image pulls in progress in parallel, if each pull takes a long time. See [issue #112044](https://github.com/kubernetes/kubernetes/issues/112044) as an example.
+
+
+### No way to limit the number of inflight image pulls
+Currently neither kubelet or containerd limits the number of inflight image pulls. On kubelet, as mentioned above, the QPS/burst limit does not take the in-progress pulls into account. On containerd, there is only [a per-image limit](https://github.com/containerd/containerd/blob/8e787543deede10f372b0e16ad5c07790fc680b6/pkg/cri/config/config.go#L299-L300), which only limits the number of parallel layer downloading for each image, but potentially allows unlimited number downloads overall.
+
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+Adding a node-level limit of parallel image pulls to kubelet. This limit will limits the maximum number of images being pulled in parallel. Any image pull request beyond the limit will be blocked until one image pull finishes.
+
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+* Prioritizing image pulls in any way.
+* Using the number of inflight image pulls as a signal to direct pod scheduling.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+* A new integer field `maxParallelImagePulls` will be added to Kubelet Configuration. `maxParallelImagePuls` is the maximum number of inflight image pulls.
+* The default value of `maxParallelImagePulls` is 0, meaning no limit on parallel image pulls, which is same as the existing behavior.
```

Comment:
I still want to promote the use of parallel image pulls to default in a future release.
-  https://github.com/kubernetes/kubernetes/issues/108405

If one day we default parallelling image pulls, what default number should maxParallelImagePuls be? Will it be changed then?

Reply:
@pacoxu i had something open https://github.com/kubernetes/kubernetes/pull/114552

Reply:
I think a default of 5 makes sense here, to match the default "--registry-qps".  Updated the KEP.